Refactor sensitivity computation API

TheMaskedTurtle opened the pull request:
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Breaking change in sensitivity computation API


**What is the current behavior?** *(You can also link to an open issue here)*
For now the computation is created by a factory that has to be specified in the platform configuration


**What is the new behavior (if this is a feature change)?**
It is made consistent with other computation modules as loadflow. The single entry point of the API is a static class which enables to find a runtime implementation of the sensitivity computation and to run it through different synchronous and asynchronous methods.


**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [x] The *Breaking Change* or *Deprecated* label has been added
- [x] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*


**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

agnesLeroy commented on sensitivity-api/src/main/java/com/powsybl/sensitivity/SensitivityComputation.java:
```diff
@@ -1,53 +1,266 @@
+
+        // if no sensitivity computation implementation name is provided through the API we look for information
+        // in platform configuration
+        String sensitivityComputationName = name != null ? name : platformConfig.getOptionalModuleConfig("sensitivity-computation")
```
It seems to me this config module is new, what parameters would be common to any sensitivity analysis provider?

agnesLeroy commented on sensitivity-api/src/main/java/com/powsybl/sensitivity/SensitivityComputation.java:
```diff
@@ -1,53 +1,266 @@
+ * @author Joris Mancini {@literal <joris.mancini at rte-france.com>}
  */
-public interface SensitivityComputation extends Versionable {
+public final class SensitivityComputation {
```
Maybe since we're breaking the API this would be an opportunity to rename SensitivityComputation into SensitivityAnalysis everywhere, since this is the standard name for it in English it seems.

TheMaskedTurtle replied:
Indeed, I did rename all the files and comments.

agnesLeroy reviewed (commented):
Thank you very much for this refactoring Joris. I have two additional requests related to it:
- could you make the corresponding changes in the sensitivity tutorial (powsybl-tutorials/sensitivity) code?
- when making changes in the tutorial code, could you also update the corresponding page on the website: github.com/powsybl/powsybl.github.io/pages/documentation/developer/tutorials/sensitivity-analysis.md
Thank you!

agnesLeroy commented on sensitivity-api/src/main/java/com/powsybl/sensitivity/SensitivityComputationProvider.java:
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Here the copyright should be 2020

TheMaskedTurtle replied:
done

agnesLeroy commented:
@TheMaskedTurtle the "find" method you reimplemented already exists, so it seems better to me to use it. I have made the changes in the commit I just pushed, please check if this is ok for you. I have also modified your unit tests to match the LoadFlow testing.

agnesLeroy reviewed (approved):
@TheMaskedTurtle could you please add some tests to increase the coverage please?

@MioRtia except for this and the copyright fixes the PR looks good to me: can I leave it to you for the merge? Thx

TheMaskedTurtle commented on sensitivity-api/src/test/java/com/powsybl/sensitivity/SensitivityComputationTest.java:
```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
```
done

TheMaskedTurtle commented on sensitivity-api/src/test/java/com/powsybl/sensitivity/mocks/AnotherSensitivityComputationProviderMock.java:
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
done

TheMaskedTurtle commented on sensitivity-api/src/test/java/com/powsybl/sensitivity/mocks/SensitivityComputationProviderMock.java:
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
done

TheMaskedTurtle commented:
@agnesLeroy I updated the code according to your comments and I also added the two following PRs : 
- For the tutorial : https://github.com/powsybl/powsybl-tutorials/pull/33
- And the website : https://github.com/powsybl/powsybl.github.io/pull/144

Thanks for your comments an review !

murgeyseb commented:
Perhaps should we also rename the artefact ? sensitivity-api -> sensitivity-analysis-api
And @mathbagu do you think we should create an extra layer in POM hierarchy just like security analysis ?

mathbagu commented on sensitivity-api/src/main/java/com/powsybl/sensitivity/json/SensitivityAnalysisParametersSerializer.java:
```diff
@@ -11,32 +11,32 @@
 
-    SensitivityComputationParametersSerializer() {
-        super(SensitivityComputationParameters.class);
+    SensitivityAnalysisParametersSerializer() {
```
Question: should we make this constructor public?

TheMaskedTurtle replied:
In the use, it is more supposed to be added to the `SensitivityAnalysisParametersJsonModule`. I think there is no "security" issue by making it public but I am not sure it will ever be used.

mathbagu commented on sensitivity-api/src/test/java/com/powsybl/sensitivity/SensitivityAnalysisTest.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Test
+    public void testAsyncNamedProvider() throws InterruptedException, ExecutionException {
+        // named provider
```
Is this comment useful (3 times)

mathbagu commented on sensitivity-api/src/test/java/com/powsybl/sensitivity/mocks/AnotherSensitivityAnalysisProviderMock.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * @author Joris Mancini {@literal <joris.mancini at rte-france.com>}
+ */
+public class AnotherSensitivityAnalysisProviderMock implements SensitivityAnalysisProvider {
```
Do you forget the `@AutoService`?

TheMaskedTurtle replied:
Yes !

TheMaskedTurtle replied:
I actually deleted this class, it was not used in the tests anymore

TheMaskedTurtle commented on loadflow/loadflow-api/src/main/java/com/powsybl/loadflow/LoadFlow.java:
```diff
@@ -96,7 +96,7 @@ public String getVersion() {
     public static Runner find(String name) {
         return new Runner(PlatformConfigNamedProvider.Finder
-                .findBackwardsCompatible(name, "load-flow", LoadFlowProvider.class,
+                .find(name, "load-flow", LoadFlowProvider.class,
```
`findBackwardsCompatible` method is declared as deprecated but indeed it might be necessary to ensure compatibility with old load flow configurations, I can go back to this method.

TheMaskedTurtle commented on sensitivity-api/pom.xml:
```diff
@@ -19,8 +19,8 @@
     </parent>
 
     <artifactId>powsybl-sensitivity-api</artifactId>
```
- What do you mean by automatic module ? 

- Then should I do as @murgeyseb suggested with a parent pom module named powsybl-sensitivity-analysis that contained a jar module powsybl-sensitivity-analaysis-api on the same model as for security analysis ?

- What would be the huge impact according to you ? Actually after sensi refactoring on FARAO side there are only two entry points of powsybl sensitivity module so adaptation will be easy

TheMaskedTurtle commented on sensitivity-api/src/main/java/com/powsybl/sensitivity/SensitivityAnalysisProvider.java:
```diff
@@ -0,0 +1,76 @@
+ * </p>
+ * @author Sebastien Murgey {@literal <sebastien.murgey at rte-france.com>}
+ */
+public interface SensitivityAnalysisProvider extends Versionable, PlatformConfigNamedProvider {
```
I thought it could be useful to have both if on implementation side only a simple sensi is implemented and not the systematic one, or only systematic and not simple. But I agree it would make the API a lot easier if there was only method in the interface....
If you still think it's not that useful I will change

TheMaskedTurtle commented on sensitivity-api/src/main/java/com/powsybl/sensitivity/SensitivityAnalysisTool.java:
```diff
@@ -1,5 +1,5 @@
-/**
- * Copyright (c) 2018, RTE (http://www.rte-france.com)
+/*
+ * Copyright (c) 2020, RTE (http://www.rte-france.com)
```
Changed to 2018-2020

TheMaskedTurtle commented on sensitivity-api/src/main/java/com/powsybl/sensitivity/SensitivityAnalysisResults.java:
```diff
@@ -35,7 +35,7 @@
  * @see SensitivityValue
  */
-public class SensitivityComputationResults {
+public class SensitivityAnalysisResults {
```
Alright

MioRtia commented on sensitivity-analysis-api/src/main/java/com/powsybl/sensitivity/analysis/json/SensitivityAnalysisParametersJsonModule.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Should the copyright be 2020? Or is it a moved file?

murgeyseb replied:
It is a moved file (renamed but moved)

MioRtia commented on sensitivity-analysis-api/src/test/java/com/powsybl/sensitivity/analysis/converter/SensitivityAnalysisResultExportersTest.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Same question

murgeyseb replied:
It is a moved file (renamed but moved)

MioRtia reviewed (approved):
Two copyright remarks but otherwise it is okay for me. Be careful, there are some conflicts with the master. Don't forget to update the migration guide!

EDIT: nevermind, I just saw it was updated already :)
